networking_ccloud/db/db_plugin.py

```diff
@@ -270,3 +296,110 @@ def remove_transit_from_network(self, context, network_id, az):
+
+    @db_api.retry_if_session_inactive()
+    def get_gateways_for_networks(self, context, network_ids, external_only=True):
+        fields = [models_v2.Subnet.network_id, models_v2.Subnet.cidr, models_v2.Subnet.gateway_ip]
+        query = context.session.query(*fields)
+        query = query.filter(models_v2.Subnet.network_id.in_(network_ids))
+        query = query.filter(models_v2.Subnet.cidr.isnot(None))
+        query = query.filter(models_v2.Subnet.gateway_ip.isnot(None))
+
+        if external_only:
+            query = query.join(extnet_models.ExternalNetwork,
+                               models_v2.Subnet.network_id == extnet_models.ExternalNetwork.network_id)
+
+        result = {}
+        for entry in query.all():
+            # we assume that the gateway is always in the network (ensured by openstack)
+            # --> we can just piece the gateway together
+            suffix = entry.cidr.split("/")[1]
+            gw_ip = f"{entry.gateway_ip}/{suffix}"
+            result.setdefault(entry.network_id, []).append(gw_ip)
+
+        # make sure the results are sorted
+        for gws in result.values():
+            gws.sort(key=lambda ip: int(ipaddress.ip_interface(ip).ip))
+
+        return result
+
+    def get_gateways_for_network(self, context, network_id, *args, **kwargs):
+        net_gws = self.get_gateways_for_networks(context, [network_id], *args, **kwargs)
+        return net_gws.get(network_id)
+
+    def get_subnet_l3_config_for_networks(self, context, network_ids):
+        """Get l3 config (cidrs, az locality) for networks, grouped by subnet pools"""
+        fields = [
+            models_v2.Subnet.subnetpool_id, models_v2.Subnet.cidr,
+            models_v2.Network.availability_zone_hints,
+        ]
+
+        query = context.session.query(*fields)
+
+        # we only get config for external networks, no reason to bother with anything else
+        # (until we maybe have DAPNET support, but that's something for the future)
+        query = query.join(extnet_models.ExternalNetwork,
+                           models_v2.Subnet.network_id == extnet_models.ExternalNetwork.network_id)
+        query = query.join(models_v2.Network,
+                           models_v2.Subnet.network_id == models_v2.Network.id)
+        query = query.filter(models_v2.Subnet.subnetpool_id.isnot(None))
+
+        # FIXME: the number of networks in the request can get quite large. would it maybe make more sense
+        #        to pre-filter these for external networks and then do the in_() to reduce query size?
+        query = query.filter(models_v2.Subnet.network_id.in_(network_ids))
+
+        # group by subnet pool
+        result = {}
+        for snp_id, cidr, net_az_hint in query.all():
+            az_hint = None
+            try:
+                if net_az_hint:
+                    net_az_hint = json.loads(net_az_hint)
+                    if len(net_az_hint) >= 1:
+                        az_hint = net_az_hint[0]
```

Comment:
Yeah, we have a `network_create_precommit()` hook that should prevent this. The `>=` is just there as a mean of "defensive programming", as I expect this to fail at some point because everything fails at some point and I'd not want the network to get regional then.